### PR TITLE
working unit tests

### PR DESCRIPTION
**Pull request by EliAndrewC**
I fixed the unit tests to work again after the big refactor.  The main change was the ``c`` refactor, which actually makes unit testing a lot easier, since we don't have to muck around with global variables as much.

While fixing the unit tests, I found several bugs, so this has already paid off :)

I'll explain the fixes inline.  @kitsuta and I can hopefully find some time within the next week or so to go over unit testing in general and walk through writing some new unit tests, etc.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -4,3 +4,4 @@ py3k-bcrypt==0.3
 stripe==1.11.0
 pytz==2014.4
 pep8==1.6.2
+pytest==2.7.2
```
Comment (EliAndrewC) [inline requirements.txt]: Sideboard ships with an older version of py.test and I wanted the later one.  I should probably update Sideboard to require this version as well.

File: setup.cfg [hunk 2]
```diff
@@ -1,4 +1,4 @@
 [pep8]
 max-line-length=999
 ignore=E121,E123,E126,E226,E24,E704,E221,E127,E128,W503,E731,E131,E711,E712
-exclude=amazon_ses.py,test_*
+exclude=amazon_ses.py
```
Comment (EliAndrewC) [inline setup.cfg]: Since our unit tests are working again, we no longer have our PEP-8 checker ignore them.

File: uber/models.py [hunk 3]
```diff
@@ -545,7 +545,7 @@ def next_badge_num(self, badge_type, old_badge_num):
 
             # Adjusts the badge number based on badges in the session
             for attendee in [m for m in chain(self.new, self.dirty) if isinstance(m, Attendee)]:
```
Comment (EliAndrewC) [inline uber/models.py]: There was a unit test for this, so it was working at some point but must have broken awhile back without us noticing.

File: uber/models.py [hunk 4]
```diff
@@ -565,52 +565,46 @@ def shift_badges(self, badge_type, badge_num, *, until=None, **direction):
                 a.badge_num += shift
 
         def change_badge(self, attendee, badge_type, badge_num=None):
```
Comment (EliAndrewC) [inline uber/models.py]: This function was heavily refactored as part of the Anthrocon branch.  On the plus side, that branch improved this by eliminating ``CUSTOM_BADGES_REALLY_ORDERED`` and replacing it with ``SHIFT_CUSTOM_BADGES``.  The downside is that several edge cases were broken by the refactor, which the unit tests revealed.

At some point we should look through this function and make sure it has literally 100% branch coverage in unit tests, since it's historically been the most brittle function in our codebase, due mainly to having so many edge cases.

File: uber/models.py [hunk 5]
```diff
@@ -1076,9 +1070,9 @@ def _badge_adjustments(self):
         # _assert_badge_lock()
 
         if self.badge_type in [c.PSEUDO_GROUP_BADGE, c.PSEUDO_DEALER_BADGE]:
```
Comment (EliAndrewC) [inline uber/models.py]: Another bug found by the unit tests.  We were changing their badge before assigning them the ribbon, at which point ``is_dealer`` would no longer be true.

File: uber/models.py [hunk 6]
```diff
@@ -1309,7 +1303,7 @@ def multiply_assigned(self):
 
     @property
     def takes_shifts(self):
```
Comment (EliAndrewC) [inline uber/models.py]: This was broken in the ``c`` refactor and we didn't notice until the unit tests found it out.  ``SHIFTLESS_DEPTS`` used to be a set and now it's a list, so this function now explicitly converts it to a set so we don't need to worry about what type of iterable it is.

File: uber/tests/conftest.py [hunk 7]
```diff
@@ -1,24 +1,49 @@
 from uber.common import *
```
Comment (EliAndrewC) [inline uber/tests/conftest.py]: There's not that much to say about the unit test fixes, but most of the changes fall into one of three categories:
- accounting for the ``c`` refactor; since we now have a global object that holds all of our config, we no longer need to mess around with module globals
- modifying or removing existing test cases because the behavior they're asserting is no longer part of Uber/RAMS (side note: is it "RAMS" or "Rams"?)
- added some default config to override whatever is in the config files so that we can write unit tests without worrying about whether ``c.ARCADE`` is actually defined, etc

I'll be happy to answer any specific questions, and we can go over unit tests in general whenever.

**Comment by binary1230** [comment]
love it.  seriously awesome. +1

File: uber/models.py (hunk 3 above)
Comment (kitsuta) [inline uber/models.py]: Is this a fix for it breaking?

File: uber/tests/models/test_badge_funcs.py [hunk 8]
```diff
@@ -1,16 +1,19 @@
+# TODO: assumes c.STAFF_BADGE and c.SUPPORTER_BADGE are preassigned badge types with configured ranges, should add this to sensible_defaults
```
Comment (kitsuta) [inline uber/tests/models/test_badge_funcs.py]: You have a TODO item here that seems to already be taken care of. :)

**Comment by kitsuta** [comment]
Eli you are the best. :D

We're one step closer to having automated testing!

File: uber/models.py (hunk 3 above)
Comment (EliAndrewC) [inline uber/models.py]: Yeah.  This is a fix for the situation where an admin manually assigns a badge number that's out of range, like if Staff badges only go up to 200 but the admin types "250" in the field.  In this case we want ``next_badge_num`` to return whatever the next available number is within the valid range, rather than returning 251.

File: uber/tests/models/test_badge_funcs.py (hunk 8 above)
Comment (EliAndrewC) [inline uber/tests/models/test_badge_funcs.py]: Whoops, good catch, I thought I had removed that when I removed my other TODOs which were taken care of.  I'll push a commit that removes this comment in a minute.

File: uber/tests/conftest.py (hunk 7 above)
Comment (earl7399) [inline uber/tests/conftest.py]: Pretty sure it's RAMS, since it's an acronym.

Comment (EliAndrewC) [inline uber/tests/conftest.py]: That makes sense, but sometimes people prefer the word-like capitalization, e.g. Perl instead of PERL, so I figured it was worth asking.
